h2o-core/src/main/java/water/Paxos.java

```diff
@@ -36,9 +36,15 @@
   // This is a packet announcing what Cloud this Node thinks is the current
   // Cloud, plus other status bits
   static synchronized int doHeartbeat( H2ONode h2o ) {
+    
+    if(h2o._heartbeat._cloud_name_hash != H2O.SELF._heartbeat._cloud_name_hash){
```

Comment:
@michalkurka This is a small additional change to this PR.

Actually we should first check if the heartbeat is coming from the correct cloud and then perform other actions. Requests from different clouds should be ignored.

Reply:
This should not be necessary. Paxos.doHeartbeat is only called from 

Paxos.doHeartbeat(SELF);

and from 

      // and if we update it here we risk dropping an update.
      ab._h2o._heartbeat = new HeartBeat().read(ab);
      if(ab._h2o._heartbeat._cloud_name_hash != H2O.SELF._heartbeat._cloud_name_hash){
        return ab;
      }
      Paxos.doHeartbeat(ab._h2o);

I suggest to remove this because this should be handled on the communication layer (wherever we still have the autobuffer it came from).

Reply:
Good catch. Removing this and merging this PR afterwards. Thanks for great help!